modules/shared/src/main/scala/org/tessellation/fsm/FSM.scala

```diff
@@ -0,0 +1,96 @@
+package org.tessellation.fsm
+
+import cats.effect.Ref
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.{Applicative, MonadThrow}
+
+import scala.util.control.NoStackTrace
+
+import derevo.cats.{eqv, show}
+import derevo.derive
+
+trait FSM[F[_], S] {
+  def state: F[S]
+
+  def set(to: S): F[Unit]
+
+  def tryModify[R](from: Set[S], through: S, to: R => S)(fn: => F[R]): F[R]
+
+  def tryModify(from: Set[S], to: S): F[Unit]
+
+
+  def tryModify[R](from: S, through: S, to: S)(fn: => F[R]): F[R] =
+    tryModify(Set(from), through, to)(fn)
+
+  def tryModify[R](from: S, through: S, to: R => S)(fn: => F[R]): F[R] = 
+    tryModify(Set(from), through, to)(fn)
+
+  def tryModify[R](from: Set[S], through: S, to: S)(fn: => F[R]): F[R] =
+    tryModify(from, through, {(_: R) => to})(fn)
+  def tryModify(from: S, to: S): F[Unit] =
+    tryModify(Set(from), to)
+}
+
+object FSM {
+
+  def make[F[_]: MonadThrow: Ref.Make, S](initial: S): F[FSM[F, S]] =
+    make(initial, _ => Applicative[F].unit)
+
+  def make[F[_]: MonadThrow: Ref.Make, S](initial: S, onChange: S => F[Unit]): F[FSM[F, S]] =
+    Ref.of[F, S](initial).map { stateR =>
+      new FSM[F, S] {
+        def state = stateR.get
+
+        def set(to: S) = stateR.set(to) >> onChange(to)
+
+        def tryModify[R](from: Set[S], through: S, to: R => S)(fn: => F[R]): F[R] = 
+          state.flatMap { initial =>
+            modify(from, through).flatMap {
+              case StateTransition.Failure => InvalidStateTransition(initial, from, through).raiseError[F, R]
```

Comment:
> Failed state modification can happen due to a concurrent update

That's not true. **Concurrent update** will not throw but will loop again.
**Parallel execution of tryModify** can fail but indeed it can fail because we cannot guarantee the expected state.

Pure FSM checks state transitions in compile-time. To recreate that, we should "expect" some initial state and try modify it. If the state is not as expected, it means either of two things: human error or other state transition in background. Both should be communicated as an error.

Let's see an example:

Pure FSM:
Node state is INITIAL -> fn changing from INITIAL to READY -> Node state is READY
Node state is INITIAL -> fn changing from READY to LEAVING  -- this will not compile. forbidden state transition

FSM with effects:

Node state is INITIAL -> transition from INTIAL to READY -> Node state is READY
Node state is INITIAL -> transition from READY to LEAVING -- this should throw because of forbidden state

Reply:
1.
>That's not true. Concurrent update will not throw but will loop again.

You need to extend your perception of a concurrent update to the one that begins not when there's a call made to `stateR.modify` but to the one that begins with obtaining the shared state value. In your case, this happens after the previous update - so just before putting a state value to the topic or a queue. It's called a `check-then-act` race condition.

An example: in the `DownloadDeamon` you're subscribed to the stream of changes to the state - once you observe `WaitingForDownload` state you start acting based on possibly stale observation - triggering download. If just before triggering download some other concurrent fiber will successfully change the state to `Leaving` then `tryModify` in the `download` function of the `Download` class will raise an error (due to a concurrent update in another fiber).

2. 
>Pure FSM checks state transitions in compile-time. To recreate that, we should "expect" some initial state and try modify it.

So the next closest thing to a compile-time check is returning `Option[R]` instead of throwing an error in F.

There's no human error if I will decide to leave the cluster, just before the download gets triggered - it's a normal flow.

Reply:
by human error I mean badly written code and expecting other state that should be at the time.
I can't agree that `Option` is closer to compile-time check as `Option` in this case is a "soft-error" but I require the FSM to be in some state. And there is no possibility FSM can be in other state when I want to execute something. If it is, it's an error and an error should be thrown

Reply:
> An example: in the DownloadDeamon you're subscribed to the stream of changes to the state - once you observe WaitingForDownload state you start acting based on possibly stale observation - triggering download. If just before triggering download some other concurrent fiber will successfully change the state to Leaving then tryModify in the download function of the Download class will raise an error - and this error will be raised solely due to a concurrent update in another fiber.

IMO this is what I'd expect from FSM.
I require state to be WaitingForDownload and I must ensure it cannot be different. If it is, it's badly designed flow of data. and if so, throw an error.

From the other hand, if I want to `leave` I don't care about anything else, that's why `set` is used instead of `modify`.

Reply:
FSM is not intended to allow for concurrent updates of state but to design state transitions in a concurrent system, while evaluating effectful computations.

Reply:
and checks for concurrent update is a `prevention` rather than `permission`

Reply:
>I require state to be WaitingForDownload and I must ensure it cannot be different. If it is, it's badly designed flow of data.

But this is how the flow of data looks currently. Also, you will never be able to ensure that the state cannot be different from expected without locking the state under some kind of a Semaphore.

>I can't agree that Option is closer to compile-time check as Option in this case is a "soft-error" but I require the FSM to be in some state.

It is closer because you force the developer to handle a potential case of invalid state transition. With an error in F, there's nothing that forces the developer to handle the invalid state transition.

Also as mentioned previously there's no way to distinguish an error in the evaluation of the fn function from a failed state transition, without matching on the class of the exception.